### DISCUSSION TITLE
userid capture for ~4 players failing

### DISCUSSION THREAD

**Discussion by TheBigBadMemeMan** [samisalreadytaken/vs_library#16] [Q&A]
hey Sam, really appreciate this library, I have been using the hell out of it. one question - I have been consistently having an issue where some of the first few players to join seemingly don't have their userid captured. https://m.twitch.tv/videos/1856821127

in my ze map cruelty squad, I am using userid for a lot of game events (namely infinite ammo but shooting your gun drains your hp), or keeping track of grenades to manually kill them to counter-balance sv_infiniteammo. 

do I need to be putting some sort of delay on when the logic_eventlistener enables for the vs.eventlistener? is there anything I can do to recapture userids? as I understand it, it should be working though based on your source code since it has checker functions... curious if you had any thoughts.

**Comment by samisalreadytaken** [comment]
Does this issue happen when there are less than 64 players?

Does the player count ever go beyond 64?

Does it happen on map transitions, map restart, round restart, in the middle of rounds; are conditions consistent?

Assuming the issue is not receiving a player from `VS.GetPlayerByUserid`, have you tried debugging it by catching the issue and doing something like this to find what the problem is:

```cs
local player = VS.GetPlayerByUserid( event.userid );

if ( !player || !player.IsValid() )
{
	if ( !player )
	{
		printf( "ERROR: Player from userid %d was not found\n", event.userid );
	}
	else if ( !player.IsValid() )
	{
		printf( "ERROR: Player from userid %d was found invalid\n", event.userid );
	}

	foreach ( pl in VS.GetAllPlayers() )
	{
		local scope = pl.GetScriptScope();
		if ( !scope )
		{
			printf( "  Player[%d] has no script scope\n", pl.entindex() );
		}
		else if ( !("userid" in scope) )
		{
			printf( "  Player[%d] has no userid\n", pl.entindex() );
		}
		else if ( scope.userid == event.userid )
		{
			printf( "  Player[%d] has userid [%d] but was not found\n", pl.entindex(), scope.userid );
		}
	}
}
```

To test more easily on a dedicated server where you don't have access to its logs, you could allow yourself to print debug information for all players to chat by checking player networkid on player_say event.

Are there any errors in the logs prefixed with `VS:`? Could be `VS::Events` or `VS::OnPlayerConnect`

Are you using the latest library version `2.43.29`? There was a related update in February.

Could I see the code?

**Reply by TheBigBadMemeMan** [reply]
the player count only goes up to 64, and it only happens for the first few folks to connect to the server (at least that's been the case when we have tested it). so it happens on map change, and it is persistent for those same players throughout the map. anyone else though, or if anyone new joins, are correctly captured. i don't have access to logs and the server admin said he didn't have any vscript errors in his logs. i'll add in a print function like you mentioned to help troubleshoot.

perhaps i have a misunderstanding with the implementation of your library, other than setting up the event listener for acquisition, i'm not doing anything extra. but just for reference, i listen for player_spawn, and the players who are not found by ```VS.GetPlayerByUserid``` are exited from the function and other functions using the same logic. When the automatic player info fails, i know that there are some additional ways to force fire a player_connect event (i think i saw a PlayerActivate function in the src for VS.Events, but i'm not sure if i should be calling that manually...)

```
::VS.ListenToGameEvent("player_spawn",function(event){
    local player = VS.GetPlayerByUserid(event.userid);
    if (player==null || !player.IsValid())return;
    local extended_player = ToExtendedPlayer(player);
    player_overlays();
    if (player.GetTeam()==2 && player.GetHealth()>7000 && mother_zombies==true){
        VS.EventQueue.AddEvent(set_zombie,0.5,[this,player]);
        VS.EventQueue.AddEvent(set_zombie,1,[this,player]);
        VS.EventQueue.AddEvent(set_zombie,3,[this,player]);        
        return;
    }
    else{
        EntFire("fade","Fade","",0,player);
        EntFireByHandle(client_command,"Command","playgamesound " + sfx_spawn_laugh.tostring(),0,player,player);
        if (!player.GetName().find("crusguy")){
            EntFireByHandle(player,"AddOutput","targetname crusguy_" + event.userid,0.01,"",""); 
        }   
    }  
    VS.EventQueue.AddEvent(set_crusguy,0.5,[this,player]);
    VS.EventQueue.AddEvent(set_crusguy,1,[this,player]);
    VS.EventQueue.AddEvent(set_crusguy,3,[this,player]); 
    player.SetMaxHealth(crusguy_hp);
    player.SetHealth(crusguy_hp);
    taxes(player);
    EntFire("loadout_checker", "Use", "", 1, player);
    give_random_weapons(player,15,false);
    EntFireByHandle(client_command,"Command","playgamesound " + sfx_health.tostring(),15,player,player);
    player.GetScriptScope().nade_count<-0;
    player.GetScriptScope().molly_count<-0;
    player.GetScriptScope().rtv<-false;
}.bindenv(this), "player_spawn_async", 0);
```

but i suppose that's my overarching question - when the userid fails to be assigned, what should i do to correct it?

**Reply by samisalreadytaken** [reply]
> server admin said he didn't have any vscript errors in his logs.

Looking for not vscript errors but regular console prints with the prefix `VS:`.

> other than setting up the event listener for acquisition, i'm not doing anything extra

This _is_ the only manual work to do.

> when the userid fails to be assigned, what should i do to correct it?

You should first determine that that is what's happening with the snippet I posted before. You can make a custom print function that besides printing, puts the messages in a buffer which you could then manually request and echo on yourself with point_clientcommand in a dedicated server. You can also do the same for vs_events by setting `VS.Events.Msg` function (right after loading the library to not miss anything) so you don't have to wait for the server logs.

There is one plausible cause for this, which I commented inside `OnPlayerActivate` function. Perhaps in ze `round_start` happens in the same frame as `player_activate` of those few players.

Either way, you could force validation with the function below. But beware of when you're calling this. Doing it on `player_spawn` might be too early, I suggest `round_start`.

```cs
function VS_ValidateAllPlayers()
{
	if ( VS.version != "2.43.29" )
		return;

	if ( !("VS_validationcache" in getroottable()) )
	{
		::VS_validationcache <- [];
		::VS_validationtime <- 0.0;
	}

	foreach( pl in VS.GetAllPlayers() )
	{
		pl.ValidateScriptScope();
		local sc = pl.GetScriptScope();
		if ( !("userid" in sc) || sc.userid == -1 )
		{
			local exit = 0;
			foreach ( i, data in getroottable()["{847D4B}"] )
			{
				if ( data && pl.entindex() == data.index+1 )
				{
					printf( "VS::ValidateAllPlayers: player%d has not yet spawned\n", pl.entindex() );
					exit = 1;
					break;
				}
			}

			if ( exit )
				continue;

			local b = 1;
			foreach( v in VS.Events.m_SV )
				if ( v == sc )
				{
					b = 0;
					break;
				};

			if (b)
			{
				printf( "VS::ValidateAllPlayers: %d\n", pl.entindex() );
				VS.Events.m_SV.append( sc.weakref() );
				EntFireByHandle( VS.Events.m_hProxy, "GenerateGameEvent", "", 0, pl, null );
			}
			else
			{
				local err = 0;
				foreach ( v in VS_validationcache )
				{
					if ( typeof v == "weakref" )
						v = v.ref();

					if ( sc == v )
					{
						printf( "VS::ValidateAllPlayers: ERROR: player%d was not validated for %f seconds!\n", pl.entindex(), Time() - VS_validationtime );

						if ( FrameTime() && (Time()-VS_validationtime) > FrameTime() )
						{
							EntFireByHandle( VS.Events.m_hProxy, "GenerateGameEvent", "", 0, pl, null );
						}

						err = 1;
						break;
					}
				}

				if ( !err )
				{
					printf( "VS::ValidateAllPlayers: player%d has not yet been validated\n", pl.entindex() );
					VS_validationcache.append( sc.weakref() );
				}
			}
		}
	}

	VS_validationtime = Time();
}
```

Note that, IIRC, it is normal for `VS.GetPlayerByUserid()` to fail once on their first player_spawn. player_spawn is fired once after player connects, then on every actual player spawn.

>```
>        if (!player.GetName().find("crusguy")){
>            EntFireByHandle(player,"AddOutput","targetname crusguy_" + event.userid,0.01,"","");
>        }
>```

Unrelated, this should be `player.GetName().find() == null` because `find` will return `0` if name was already set. You're also passing empty strings to activator and caller parameters. It's harmless, but weird. You can just skip them.

I also suggest setting the delay to `0.05` or `FrameTime() * 2` to avoid certain issues, or alternatively you could set name without any delay (`__KeyValueFromString()` or `VS.SetName()`) _before_ `ToExtendedPlayer()` call. The very first `ToExtendedPlayer()` changes player name for one server frame. It is then reverted after `tick_interval + 0.001` second (`0.016` on 64 tick).

**Reply by samisalreadytaken** [reply]
Any updates?
